flink-runtime/src/main/java/org/apache/flink/runtime/zookeeper/ZooKeeperStateHandleStore.java

```diff
@@ -154,54 +158,75 @@ public ZooKeeperStateHandleStore(
             return storeHandle;
         } catch (Exception e) {
             if (indicatesPossiblyInconsistentState(e)) {
                 throw new PossibleInconsistentStateException(e);
             }
-
-            // in any other failure case: discard the state
+            // In case of any other failure, discard the state and rethrow the exception.
             storeHandle.discardState();
-
-            // We wrap the exception here so that it could be caught in DefaultJobGraphStore
-            throw ExceptionUtils.findThrowable(e, KeeperException.NodeExistsException.class)
-                    .map(
-                            nee ->
-                                    new AlreadyExistException(
-                                            "ZooKeeper node " + path + " already exists.", nee))
-                    .orElseThrow(() -> e);
+            throw e;
         }
     }
 
     // this method is provided for the sole purpose of easier testing
     @VisibleForTesting
     protected void writeStoreHandleTransactionally(String path, byte[] serializedStoreHandle)
             throws Exception {
-        // Write state handle (not the actual state) to ZooKeeper. This is expected to be
-        // smaller than the state itself. This level of indirection makes sure that data in
-        // ZooKeeper is small, because ZooKeeper is designed for data in the KB range, but
-        // the state can be larger.
-        // Create the lock node in a transaction with the actual state node. That way we can
+        // Write state handle (not the actual state) to ZooKeeper. This is expected to be smaller
+        // than the state itself. This level of indirection makes sure that data in ZooKeeper is
+        // small, because ZooKeeper is designed for data in the KB range, but the state can be
+        // larger. Create the lock node in a transaction with the actual state node. That way we can
         // prevent race conditions with a concurrent delete operation.
-        client.inTransaction()
-                .create()
-                .withMode(CreateMode.PERSISTENT)
-                .forPath(path, serializedStoreHandle)
-                .and()
-                .create()
-                .withMode(CreateMode.EPHEMERAL)
-                .forPath(getLockPath(path))
-                .and()
-                .commit();
+        while (true) {
+            try {
+                client.inTransaction()
+                        .create()
+                        .withMode(CreateMode.PERSISTENT)
+                        .forPath(path, serializedStoreHandle)
+                        .and()
+                        .create()
+                        .withMode(CreateMode.EPHEMERAL)
+                        .forPath(getLockPath(path))
+                        .and()
+                        .commit();
+                break;
+            } catch (KeeperException.NodeExistsException e) {
```

Comment:
We could also read the contents of `path` to see whether we have written the `serializedStoreHandle`. If this is the case, then the call succeeded and otherwise we can throw the `NodeExistsException`.

Reply:
We now check whether node is empty before executing `writeStoreHandleTransactionally()`, so this is hopefully covered by that. We can IMO trust the transaction mechanism here (assuming there are no concurrent modifications to the znode).